### DISCUSSION TITLE
Payload 1.6 Megathread

### DISCUSSION THREAD

**Discussion by denolfe** [payloadcms/payload#1992] [General]
Version 1.6 has been released 🎉

This is a very large release that comes with some 🚨 BREAKING CHANGES 🚨 , so please read the [Changelog](https://github.com/payloadcms/payload/releases/tag/v1.6.1) carefully.

We try our best to avoid breaking changes, but we decided this was the right time to do it as it paves the way for greater things in the future like ongoing TypeScript niceties, serverless support, and additional database support 😎

If you run into issues after reading the changelog, please reply in this thread, so we can assist.

**Comment by STR1234** [comment]
After the update and going through the Changelog printSchema.js throws:
`const directives = schema.getDirectives().filter(directiveFilter);`

`TypeError: Cannot read properties of undefined (reading 'getDirectives')`
whenever generate:graphQLSchema is invoked.
Any ideas? :) What have I missed? 🤔

**Reply by denolfe** [reply]
Good catch! Looking into this now.

**Reply by denolfe** [reply]
@STR1234 Fixed in 1.6.3

**Reply by STR1234** [reply]
Yeeees indeed 👍  Thanks for the fast response :)

**Comment by vstarush** [comment]
Hello! For non-Typescript project in version 1.6.3 the payload.config.js and all collections are throwing an error:
"SyntaxError: Cannot use import statement outside a module."

Not sure if this is because the payload.init became async or because of any other changes.
A fix like adding 'type: "module"' in package.json doesn't work cause it breaks other things and changing payload.config from js to .mjs doesn't either.

Would appreciate on any ideas on how to fix this, if this is at all possible. Thank you! :)

**Reply by JarrodMFlesch** [reply]
Hey @vstarush! You can still use JS without using TS, there are just a couple things that you will need to update. The changelog has a link out to a PR that explains the few steps that are needed to get your JS project working. Here is the [PR](https://github.com/payloadcms/payload/pull/1847) with the steps you will need to follow.

Let us know if you run into anything 👍

**Comment by fabioquarantini** [comment]
Does "retainDeleted" no longer exist?

**Reply by jmikrut** [reply]
From the start, that property was actually incorrectly documented and it never did anything. Versions **_must_** be deleted alongside of their parent documents being deleted, but in the future, we can build this feature if you need it. What do you think?

**Reply by fabioquarantini** [reply]
Ok, the "Trash" function would be ideal. (Like WordPress)

**Comment by thefrana** [comment]
There was a mention of Vercel deployments. Is there some guide on how to deploy Payload to it?

**Reply by jmikrut** [reply]
Not yet - but keep an eye out 😈

**Comment by jakehopking** [comment]
Hi - I've upgraded to the 1.6.6 and all records in my collections (in the admin ui) are missing.

I've followed the instructions here: https://github.com/payloadcms/payload/pull/1847

So I can confirm that dev runs using `ts-node` despite loading `.js` files which contain `require`.

I've tried running your migration script, however it fails with this error:
```shell
❯ PAYLOAD_CONFIG_PATH=payload.config.ts npx ts-node -T scripts/payloadMigrateVersions.ts
[10:52:31] INFO (payload): Connected to Mongo server successfully!
[10:52:31] INFO (payload): Starting Payload...
/Users/jake/Development/personal/elkli-and-hart/payload.config.ts:1
import path from 'path';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

### Type = modules

If I set `"type": "module"` in `package.json` I then get this error:

```shell
❯ PAYLOAD_CONFIG_PATH=payload.config.ts npx ts-node -T scripts/payloadMigrateVersions.ts
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/jake/Development/personal/elkli-and-hart/scripts/payloadMigrateVersions.ts
```

... and another side effect of the `type: modules` is dev breaks cos I use `require` within `dev.js` and `server.js`.

### `--esm`

Running with the `--esm` flag obviously doesn't work either:

```shell
❯ PAYLOAD_CONFIG_PATH=payload.config.ts npx ts-node --esm -T scripts/payloadMigrateVersions.ts
ReferenceError: require is not defined in ES module scope, you can use import instead
```

### `--swc`

I have this already in my `tsconfig.json`

```json
"ts-node": {
  "swc": true
}
```

But trying to run with this flag just to be sure, still results in a fail (albeit slightly different):

```shell
❯ PAYLOAD_CONFIG_PATH=payload.config.ts npx ts-node --swc -T scripts/payloadMigrateVersions.ts
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/jake/Development/personal/elkli-and-hart/scripts/payloadMigrateVersions.ts
```


#### My payload config is typescript.

Here it is for reference:

```ts
import path from 'path';
import dotenv from 'dotenv';
import { buildConfig } from 'payload/config';
import {
  Category,
  Collection,
  Essay,
  Media,
  MediaPhotos,
  Page,
  Price,
  Product,
  ReusableContent,
  Tag,
  User,
} from './collections';
import { SiteNavigation, Footer } from './globals';

dotenv.config();

export default buildConfig({
  serverURL: process.env.PAYLOAD_PUBLIC_SERVER_URL,
  typescript: {
    outputFile: path.resolve(__dirname, './generated-types.ts'),
  },
  // admin: {
  //   user: 'admins',
  // },
  graphQL: {
    disable: true,
  },
  collections: [
    Product,
    Essay,
    Collection,
    Page,
    Media,
    MediaPhotos,
    ReusableContent,
    Price,
    Category,
    Tag,
    User,
  ],
  globals: [Footer, SiteNavigation],
  debug: true,
  // optional customization of routes
  routes: {
    api: '/payload',
    // admin: '/admin',
    // graphQL: '/graphql',
    // graphQLPlayground: '/graphql-playground',
  },
});

```

This release feels rushed and I'm stuck with broken admin dashboard again. It was only a month ago that a payload update broke versions completely for me, and we had to work together to get that fixed. Why are we here again?

Please can you provide me with a migration script that works?

**Reply by jakehopking** [reply]
For now commenting out this in the collections with versions allows the admin ui to display all items.

```ts
  // versions: {
  //   maxPerDoc: 5,
  //   drafts: {
  //     autosave: false,
  //   },
  // },
```

And dropping all `_[collectionName]_versions` within mongodb, and then uncommenting that above code, the admin ui again doesn't show my records  #sigh

---

I've restored my db and have left the commented out code in all collections which have versions enabled.

**Reply by jmikrut** [reply]
Hey @jakehopking I totally hear you. Honestly a big part of the changes in 1.6.0 were _**as a result**_ of the versions issues we were experiencing prior. This update intends to once and for all dramatically simplify the underlying logic for versions and make it so they are significantly more bulletproof moving forward.

I don't think that this update was rushed though. It's certainly unfortunate that it came with breaking changes / migrations necessary, but this is something that needed to happen either immediately, or at some point in the future. That said, we decided as a team to collectively rip off the bandaid and provide as much help as possible to move forward with the new and simplified versions logic. And we've tested / used the migration script ourselves while these changes were released as `canary` on 5+ large production Payload instances. 

That all said, I am confident we can get this script to work for you. Here are some quick thoughts:

- We've tested with TS configs, and it's interesting that your TS config itself is causing the error, but the script is not. Maybe try converting your script to ES imports. There could be something with your ts-node or typescript version locally that sees a `require()` call, and then treats all scripts as that moving forward (thus failing when it sees an `import`.)
- You should not need to put `type: module` in your package.json file
- You should try and compare / contrast your `tsconfig.json` to one that is generated with `create-payload-app`. 
- What is your `tsconfig` `compilerOptions.target` set to? 

Here is a `tsconfig` that works with the script:

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "strict": false,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "outDir": "./dist",
    "rootDir": "./src",
    "jsx": "react",
    "paths": {
      "payload/generated-types": [
        "./src/payload-types.ts",
      ],
    }
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules",
    "dist",
    "build",
  ],
  "ts-node": {
    "transpileOnly": true,
    "swc": true,
  }
}
```

Note that `esModuleInterop` is set to `true`. Do you have that set to `true` in your `tsconfig`?

Sorry about the hassle here! We will get through it!

**Reply by jakehopking** [reply]
Thanks @jmikrut - I can now successfully run the script... however we're not home and dry yet (see from below image).

> As a note, whenever I run dev, NextJS updates the `tsconfig` with the following additions:

<img width="726" alt="Screenshot 2023-02-08 at 19 42 14" src="https://user-images.githubusercontent.com/3432643/217635160-ee4168ad-4b92-4e0c-8d1b-d8f29f73c584.png">

And whenever you try to run the migration script after Next updates the config, we run into the issues I outlined in the OP. 🤦 

### Records still missing

Even though I have run the script and saw this:
```shell
❯ PAYLOAD_CONFIG_PATH=payload.config.ts npx ts-node -T scripts/payloadMigrateVersions.ts
[19:36:07] INFO (payload): Connected to Mongo server successfully!
[19:36:07] INFO (payload): Starting Payload...
Done!
```

I'm missing records from the admin ui (yes, I've uncommented my versions config in all the collections).

There should be more than a single page for instance:

Any ideas?

<img width="891" alt="Screenshot 2023-02-08 at 19 47 57" src="https://user-images.githubusercontent.com/3432643/217635945-c2129ef7-d767-47bb-a762-529042c4d564.png">

**Reply by jmikrut** [reply]
OK getting somewhere. So you're using Next + Payload in a single repo and Next is adjusting your tsconfig for you—did you try to update the script itself to use `import` instead of `require`? Maybe if you do that, you wouldn't need to make any adjustments at all to your `tsconfig`.

Which `tsconfig` update made this work for you, out of curiosity?

Now, onto what we need to diagnose next.

Basically, the changes that we've released in 1.6.0 make it so that when you run the `find` operation with `?draft=true`, we _**only query on the versions collection**_ - and not the main collection. This is for performance, simplicity, database compatibility, etc. 

The script should go through your pages, and find any pages that _**do not have at least one corresponding version present.**_ If there are no versions for a document in your main collection, then the script will create a version corresponding to the content of the document in your main collection. 

The output of the script tells me that it is not creating any versions for you. Note that you will need to re-enable drafts on your collections for the script to work. I know you commented out drafts temporarily, but the script only runs against collections that have drafts enabled.

Basically, we need to figure out why the script is not creating corresponding versions. Working through the script contents should reveal the answer, but it might be due to the above. If not that, then something simlar.

Do you have any versions in your _versions collection?

**Reply by jakehopking** [reply]
Thanks for the info James, and makes sense.

I did indeed try to change the script to use `import` rather than `require` but it ultimately made no difference and generated other errors, which I didn't share cos then the variables had increased :)

You are correct, I forgot to uncomment my versions config **before** running the script. With that done, we have a successful log of changes:

<img width="764" alt="Screenshot 2023-02-08 at 20 15 56" src="https://user-images.githubusercontent.com/3432643/217642407-30a99d83-19d0-4c5b-b617-c8a0a592fe37.png">

Thanks for your help 👍 

P.S. This config worked for the migration (essentially the one you shared), although not so much for Next :/

```ts
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "strict": false,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "outDir": "./dist",
    "rootDir": "./",
    "jsx": "react",
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules",
    "dist",
    "build",
  ],
  "ts-node": {
    "transpileOnly": true,
    "swc": true,
  }
}
```

**Reply by jmikrut** [reply]
🔥 thank you for hanging in there!

I think versions are going to be much more reliable moving forward. I've been totally expecting people to have trouble and have been sad that we've had to introduce a breaking change, but I do think that in the end it will be very worth it. Thanks again.

**Reply by JarrodMFlesch** [reply]
🚨 @jakehopking please update to v1.6.9 when you can. There was a bug with versions that surfaced in the 1.6.0 release because we started setting the `max` versions to 100 by default. This was a bug all along but since max was not being set by default, the bug would have only occurred if a custom max value was set on a versioned collection.

More info here: https://github.com/payloadcms/payload/pull/2058

**Comment by unonweb** [comment]
If anyone else is running in trouble with version 1.6 because of a javascript installation:
_Just choose the typescript installation_ (`npx create-payload-app`, then choose "typescript").

After updating to 1.6 my Payload broke as predicted due to [this](https://github.com/payloadcms/payload/pull/1847).
I followed the instructions behind that link without success ("typescript" module was missing although I had installed ts-node).
Then I tried a new installation with `npx create-payload-app` choosing javascript. Not working.
But when I chose Typescript everything was working again (and I can still go on writing js).

**Reply by denolfe** [reply]
Hey @unonweb , in the latest `create-payload-app` we decided to drop all of the JS templates. It looks like you may have a cached version. That being said, you absolutely can still use JavaScript. Can you provide more detail on what specifically isn't working after following those instructions?

**Reply by unonweb** [reply]
Sorry, in the end this was my error:
> "typescript" module was missing although I had installed ts-node

`npm i typescript` in addition to `npm i ts-node` resolved it.
